### ISSUE TITLE
ImageRepository manifests ignoring spec.secretRef changes 

### ISSUE THREAD

**Issue by stvnksslr** [fluxcd/flux2#2286]
### Describe the bug

We noticed this issue after updating to v0.25.1 and the issue is not currently effecting one of our other clusters that is on v0.24.1

when making changes to our image repository manifests we noticed that despite the reconciliation passing without issue the spec.secretRef field was not effected example.

Git Version
```
apiVersion: image.toolkit.fluxcd.io/v1beta1
kind: ImageRepository
metadata:
  name: some-app
  namespace: flux-system
spec:
  image: <ECR_URL>/some-app
  interval: 5m
```

Cluster Version
```
apiVersion: image.toolkit.fluxcd.io/v1beta1
kind: ImageRepository
metadata:
  name: some-app
  namespace: flux-system
spec:
  image: <ECR_URL>/some-app
  interval: 5m
  secretRef:
    name: ecr-credentials
```

### Steps to reproduce

1. add a spec.secretRef section to an existing ImageRepository manifest
2. commit to git
3. watch reconciliation pass successfully 
4. remove field 
5. watch reconciliation pass successfully 
6. see that spec.secretRef has not been removed

### Expected behavior

I expect that when removing the spec.secretRef for the sync process to remove it on the cluster as well or error if there is a reason it cannot be edited/applied.

### OS / Distro

N/A

### Flux version

v.0.25.1

### Flux check

► checking prerequisites
✔ Kubernetes 1.21.5 >=1.19.0-0
► checking controllers
✔ helm-controller: deployment ready
► ghcr.io/fluxcd/helm-controller:v0.15.0
✔ image-automation-controller: deployment ready
► ghcr.io/fluxcd/image-automation-controller:v0.19.0
✔ image-reflector-controller: deployment ready
► ghcr.io/fluxcd/image-reflector-controller:v0.15.0
✔ kustomize-controller: deployment ready
► ghcr.io/fluxcd/kustomize-controller:v0.19.0
✔ notification-controller: deployment ready
► ghcr.io/fluxcd/notification-controller:v0.20.1
✔ source-controller: deployment ready
► ghcr.io/fluxcd/source-controller:v0.20.1
✔ all checks passed

### Git provider

gitlab

### Container Registry provider

ECR

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by kingdonb** [comment]
I have done a test to try to confirm this report, and I was able to confirm so far the issue seems to exist at least as far back as Flux 0.24.1

In my case I removed a `secretRef` from a `GitRepository` in a way that should have definitely made it invalid, but it synced OK. When I deleted the `GitRepository` from the cluster and let it sync again from an upstream Kustomization, I found it was definitely invalid:

* e3d8299e98fc545b28e6f256e189d6a9338d8f24
* cdc97dfdea7f5b8d9ae5262ade98b86d16e62a04

Applied on a fresh cluster with 0.24.1 (or even just deleting the earlier GitRepository first) the first of these commits, I would have expected to remove the fields which Flux applied in earlier versions, but they are not removed. The second commit is definitely invalid according to the `--dry-run=server` apply, but it gets accepted without issue and the Kustomization that applies it says:

```
flux-system           	True 	Applied revision: staging/cdc97dfdea7f5b8d9ae5262ade98b86d16e62a04   	staging/cdc97dfdea7f5b8d9ae5262ade98b86d16e62a04   	False
```

I'm going to start checking older versions of Flux since around 0.18 since that's when server side apply was first used and see if they all mirror this behavior, or if I can narrow down by bisecting which change actually introduced this behavior, since it does not seem like correct behavior and I'm assuming this information will be helpful now that I've reproduced an issue.

Thanks for your report @stvnksslr I'll let you know what I find out 👍

**Comment by kingdonb** [comment]
There is definitely a regression between 0.18.3 and 0.24.1, I'm working on narrowing it down now.

* [02833d833eb3c51169c81c7be30eb222a6c49532](https://github.com/kingdonb/bootstrap-repo/commit/02833d833eb3c51169c81c7be30eb222a6c49532)

In Flux 0.18.3, this commit definitely removes the `secretRef` from the GitRepository as I'd expect it to.

**Comment by kingdonb** [comment]
I found an error in my testing approach, something has changed and I no longer seem to have a repro for this issue now that I corrected my error. Even on the latest version, where I thought I was able to reproduce this issue easily on the first try, now it does not seem to be throwing a problem at all.

The GitRepository updates and removes its `secretRef`, then errors out predictably as you'd expect any SSH GitRepos to fail if they did not have any SSH key attached.

I'll spend some more time on it to see if I can see what I was seeing again, since it would be very helpful to have a clean repro for some of the server side apply issues we've seen reported for a while now.

**Comment by kingdonb** [comment]
Hmm, it seems the issue reproduces cleanly on a new v0.25.1 installation, as in fresh new cluster which has only just had v0.25.1 installed on it, without upgrading from any earlier version. If I create a GitRepository or some other resource with a `secretRef` (or I presume any other optional field that can be omitted, as I have no reason to think this issue is specific to `secretRef`), then commit a new version of the manifest which is missing the `secretRef`, I do not see the field disappear from the cluster-based representation of the resource after the latest manifest has been applied by Flux.

If I upgrade from one version to the next, the issue does not reproduce reliably anymore. I am still trying to isolate the first version that had this issue. It takes some time because I have to destroy the entire cluster and try again with each suspect version on a fresh cluster, almost seems as though it matters which version of Flux is upgraded to which version of Flux.

**Comment by kingdonb** [comment]
v0.23.0 and prior does not seem to have this bug. I can bootstrap Flux fresh on a new cluster with this version, then remove the secretRef, then edit the live resource on the cluster and see the `spec.secretRef` section has been removed as soon as Flux synced. Later versions fail to remove the secret ref.

The issue that is described here is surfacing for me seems to be reproduced in v0.24.0 and on, all suffer from the issue described here when installed on a fresh cluster. But for some reasons, couldn't be reproduced in every cluster.

**Comment by stefanprodan** [comment]
This is related to https://github.com/fluxcd/kustomize-controller/issues/486 you can't delete fields added by kubectl. You need to clear the managed fields first.

**Comment by yebyen** [comment]
The repro was done on a fresh K8s 1.23.1 cluster which has never been touched by kubectl - with only Flux 0.25.1 ever installed on it - I'll try it again today to confirm for sure, but this doesn't jibe with the managed fields explanation if it's supposed to have been fixed in 0.25

**Comment by stefanprodan** [comment]
> if it's supposed to have been fixed in 0.25

The managed fields issue is still opened, but we’ve updated the Kubernetes packages which could improve the situation.

**Comment by kingdonb** [comment]
Wouldn't users only experience this bug if they were upgrading from an older version of Flux (earlier than SSA) to a newer one? I'm not understanding how I would have come into any fields managed by kubectl, when this resource was added by Flux and synced through the gitrepo on the latest version of Flux, and this cluster has never been touched by kubectl except for completely unrelated kubectl apply to add the `sops-gpg` after the `flux-system` kustomization and namespace.

I have been able to confirm again that K8s 1.23.1 with Flux 0.25.1, never operated on any earlier version, apparently has this issue and the managedFields patch approach to solving it offered above has no affect, as it appears the resource has no managedFields to patch and remove when it is opened with `kubectl edit --show-managed-fields`.

**Comment by stefanprodan** [comment]
Can you post here the object with the managed fields when it gets in that state where removing the spec.secretRef from Git has no effect?

**Comment by stefanprodan** [comment]
I found the issue, what happens is that image-reflector-controller acts like kubectl when it adds finalizers and becomes the manager of all the fields inside `.spec`, this means the Kubernetes API will not allow kustomize-controller to delete any fields from the spec because some other controller manages them as well.

**Comment by stefanprodan** [comment]
To solve this issue we need to replace everywhere (in all our controllers) `client.Update` with `client.Patch` when we add finalizers, e.g. https://github.com/fluxcd/source-controller/blob/main/controllers/gitrepository_controller.go#L104

**Comment by kingdonb** [comment]
I can submit PRs with this simple change for all repos to close this issue, 🌮 

That's a very obscure problem and I'm glad you could see it. I should be able to confirm pretty quickly if this change in source-controller fixes the issue I'm finding.

**Comment by kingdonb** [comment]
Thank you, I'm able to confirm the new `source-controller` doesn't take control of the resources in the same way with the finalizer changes, and a fresh cluster bootstrapped with this better `source-controller` removes the `secretRef` when it is taken out of the source manifest, 🎉 there are some things to figure out to close the book on this,

I'm going to be looking at:

* fluxcd/pkg#209

to understand whatever I need to know to tell users how they can get out of this pickle if they've encountered it already.

**Comment by kingdonb** [comment]
It looks like fluxcd/pkg#209 and fluxcd/source-controller#542 together are not quite yet enough to liberate resources which were once "update"d to set their `finalizers` by source controller, I did the upgrade test and used upgraded versions of both controllers. @stefanprodan FYI

I deployed the `rc-408a889a` kustomize controller from fluxcd/kustomize-controller#527 and a `source-controller` with the latest main branch including 542, and the issue remains present on the cluster if it was bootstrapped between 0.24.0 and 0.25.2, I'm still not sure how to clear the ownership from the resource so that Flux can manage it again.

We will need something else to tell users who hit this issue, or ideally something we can bake into the next release so they don't have to fix it on every given resource for themselves. (We are likely to get some questions from users after they merge this fix, however we can accomplish that, as latent changes that may have been stuck unapplied are resolved.)

**Comment by stefanprodan** [comment]
We could write a `flux fix` command that does a proper cleanup of all the toolkit resources in a cluster. For all the other resources applied with kustomize-controller (back when it used kubectl), PR 527 should fix it.

**Comment by stefanprodan** [comment]
> Is it only finalizer Update that we need to worry about, or all instances of Update that modify any resources at all?

All updates besides the one that finalizes a resource before deleting.

**Comment by jmriebold** [comment]
Hey all, looks like this is fixed as of v0.26.0. Upgraded this morning and then removed the fields and everything worked as expected. Thanks @kingdonb and @stefanprodan!

**Comment by stefanprodan** [comment]
Great news! Thanks @jmriebold for confirming the fix.
